### DISCUSSION TITLE
Bug when compiling prisma using custom output

### DISCUSSION THREAD

**Discussion by dinkelspiel** [prisma/prisma#23818] [Q&A]
### Bug description

When I try to compile my nextjs project with after my prisma codegen I get the error.

```Dockerfile
29.01 qrtag:build: ./prisma/client/runtime/edge-esm.js:1:1
29.01 qrtag:build: Type error: Declaration emit for this file requires using private name 'i'. An explicit type annotation may unblock declaration emit.
```

https://cdn.discordapp.com/attachments/1225878190891864104/1226937916312518707/image.png?ex=662695fa&is=661420fa&hm=6e04892eb54415c469c92fbe5b131f9abfcb1fbbf0f93037fd438f17eb787b21&

The project is running inside a turborepo and i'm trying to create a docker image using the following dockerfile.

```
# Builder
FROM node:alpine AS builder
RUN apk add --no-cache libc6-compat
RUN apk update

WORKDIR /app
RUN yarn global add turbo
COPY . .
RUN turbo prune --scope=qrtag --docker


# Installer
FROM node:alpine AS installer
RUN apk add --no-cache libc6-compat
RUN apk update

WORKDIR /app
RUN yarn global add pnpm
RUN yarn global add turbo

COPY /apps/qrtag/prisma .
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
COPY --from=builder /app/out/pnpm-workspace.yaml ./pnpm-workspace.yaml
RUN pnpm fetch
RUN pnpm install --offline
RUN pnpm add prisma@5.12.1 --workspace-root
RUN pnpm add @prisma/client@5.12.1 --workspace-root
RUN pnpx prisma generate --schema=./schema.prisma

COPY --from=builder /app/out/full/ .
COPY turbo.json turbo.json

RUN turbo build --filter=qrtag...

# Runner
FROM node:alpine AS runner
WORKDIR /app

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs

COPY --from=installer /app/apps/qrtag/next.config.js .
COPY --from=installer /app/apps/qrtag/package.json .

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=installer --chown=nextjs:nodejs /app/apps/qrtag/.next/standalone ./
COPY --from=installer --chown=nextjs:nodejs /app/apps/qrtag/.next/static ./apps/qrtag/.next/static
# COPY --from=installer --chown=nextjs:nodejs /app/apps/appname/public ./apps/appname/public

CMD node apps/qrtag/server.js
```

### How to reproduce

No idea

### Expected behavior

Expected behavior is that the compilation works and the NextJS compiles without type errors.

### Prisma information
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../prisma/client"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model EventTag {
  id        Int       @id @default(autoincrement()) @db.UnsignedInt
  userId    Int       @db.UnsignedInt
  eventId   Int       @db.UnsignedInt
  targetId  Int       @db.UnsignedInt
  timestamp DateTime  @default(now()) @db.DateTime(0)
  target    User      @relation("event_tags_target_idTousers", fields: [targetId], references: [id], onUpdate: Restrict, map: "event_tags_target_id_foreign")
  eventUser EventUser @relation(fields: [userId, eventId], references: [userId, eventId])
  user      User      @relation(fields: [userId], references: [id])
  event     Event     @relation(fields: [eventId], references: [id])

  @@index([targetId], map: "event_tags_target_id_foreign")
}

model EventUser {
  id       Int        @id @default(autoincrement()) @db.UnsignedInt
  userId   Int        @db.UnsignedInt
  eventId  Int        @db.UnsignedInt
  isAlive  Boolean    @default(true)
  secret   String     @db.VarChar(255)
  targetId Int?       @db.UnsignedInt
  event    Event      @relation(fields: [eventId], references: [id], onUpdate: Restrict, map: "event_users_event_id_foreign")
  target   User?      @relation("event_users_target_idTousers", fields: [targetId], references: [id], onUpdate: Restrict, map: "event_users_target_id_foreign")
  user     User       @relation("event_users_user_idTousers", fields: [userId], references: [id], onUpdate: Restrict, map: "event_users_user_id_foreign")
  tags     EventTag[]

  @@unique([userId, eventId])
  @@index([eventId], map: "event_users_event_id_foreign")
  @@index([targetId], map: "event_users_target_id_foreign")
  @@index([userId], map: "event_users_user_id_foreign")
}

model Event {
  id              Int         @id @default(autoincrement()) @db.UnsignedInt
  name            String      @db.VarChar(255)
  startDate       DateTime    @db.DateTime(0)
  endDate         DateTime    @db.DateTime(0)
  targetsAssigned Boolean     @default(false)
  winnerId        Int?        @db.UnsignedInt
  winner          User?       @relation(fields: [winnerId], references: [id])
  EventUser       EventUser[]
  EventTag        EventTag[]
}

model User {
  id              Int         @id @default(autoincrement()) @db.UnsignedInt
  username        String      @unique @db.VarChar(255)
  displayName     String      @db.VarChar(255)
  class           String      @db.VarChar(6)
  createdAt       DateTime    @default(now()) @db.DateTime(0)
  updatedAt       DateTime    @default(now()) @db.DateTime(0)
  eventTargetTags EventTag[]  @relation("event_tags_target_idTousers")
  eventTargets    EventUser[] @relation("event_users_target_idTousers")
  eventUsers      EventUser[] @relation("event_users_user_idTousers")
  userAdmin       UserAdmin?
  Event           Event[]
  EventTag        EventTag[]
}

model UserAdmin {
  id        Int      @id @default(autoincrement()) @db.UnsignedInt
  userId    Int      @unique @db.UnsignedInt
  user      User     @relation(fields: [userId], references: [id])
  createdAt DateTime @default(now()) @db.DateTime(0)
  updatedAt DateTime @default(now()) @db.DateTime(0)
}
```

### Environment & setup

- OS: Linux  6.8.2-arch2-1 
- Database: MySQL
- Node.js version: v20.12.1

### Prisma Version

prisma                  : 5.11.0
@prisma/client          : 5.11.0
Computed binaryTarget   : debian-openssl-3.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v20.12.1
Query Engine (Node-API) : libquery-engine efd2449663b3d73d637ea1fd226bafbcf45b3102 (at ../../node_modules/.pnpm/@prisma+engines@5.11.0/node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli efd2449663b3d73d637ea1fd226bafbcf45b3102 (at ../../node_modules/.pnpm/@prisma+engines@5.11.0/node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.11.0-15.efd2449663b3d73d637ea1fd226bafbcf45b3102
Default Engines Hash    : efd2449663b3d73d637ea1fd226bafbcf45b3102
Studio                  : 0.499.0

**Comment by ludralph** [comment]
Hi @dinkelspiel 👋 

What version of Typescript are you using within the Turbo repo? Also, how are you importing the edge client? If you use the default output directory, do you still experience this error?

**Comment by WillKirkmanM** [comment]
Hey @dinkelspiel. I have encountered the same issue and this sequence of steps allowed a successful NextJS build without type errors 🥳. 
TL;DR: Create a script in `package.json` that generates the database with `$ prisma generate` (link it in `turbo.json`). Then, before the build command, run the `turbo run generate --scope <YOUR APP>` (where app is the package.json given name. **NOT** the folder name).

Below is my example, the docker file is taken from [the handbook](https://turbo.build/repo/docs/handbook/deploying-with-docker):

```docker
FROM node:21-alpine AS base
 
FROM base AS builder
RUN apk add --no-cache libc6-compat
RUN apk update
# Set working directory
WORKDIR /app
RUN yarn global add turbo
COPY . .
RUN turbo prune music --docker
 
# Add lockfile and package.json's of isolated subworkspace
FROM base AS installer
RUN apk add --no-cache libc6-compat
# RUN apk add --update --no-cache openssl1.1-compat
RUN apk add --update --no-cache openssl
RUN apk update
WORKDIR /app
 
# First install the dependencies (as they change less often)
# COPY .gitignore .gitignore
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/yarn.lock ./yarn.lock

RUN yarn install

# Build the project
COPY --from=builder /app/out/full/ .

RUN yarn turbo run generate --scope music
RUN yarn turbo run build --filter=music
 
FROM base AS runner
WORKDIR /app
 
# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs
 
COPY --from=installer /app/apps/web/next.config.mjs .
COPY --from=installer /app/apps/web/package.json .
 
# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=installer --chown=nextjs:nodejs /app/apps/web/.next/standalone ./
COPY --from=installer --chown=nextjs:nodejs /app/apps/web/.next/static ./apps/web/.next/static
COPY --from=installer --chown=nextjs:nodejs /app/apps/web/public ./apps/web/public

CMD node apps/web/server.js
```

NOTE:
```
. . .
# First install the dependencies (as they change less often)
# COPY .gitignore .gitignore
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/yarn.lock ./yarn.lock

RUN yarn install

# Build the project
COPY --from=builder /app/out/full/ .

RUN yarn turbo run generate --scope music
RUN yarn turbo run build --filter=music
 
FROM base AS runner
WORKDIR /app
. . .
```

My `turbo.json`:
```
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": ["**/.env.*local"],
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [".next/**", "!.next/cache/**"]
    },
    "lint": {
      "dependsOn": ["^lint"]
    },
    "dev": {
      "cache": false,
      "persistent": true
    },
    "generate": {
      "dependsOn": [],
      "outputs": ["@prisma/client"]
    }
  }
}
```
Root `package.json`
```
{
  "name": "music",
  "private": true,
  "scripts": {
    "build": "turbo build",
    "dev": "turbo dev",
    "lint": "turbo lint",
    "format": "prettier --write \"**/*.{ts,tsx,md}\"",

    "generate": "prisma generate",
    "push": "prisma db push",
    "studio": "npx prisma studio",
    "migrate": "prisma migrate dev"
  },
  "devDependencies": {
    "prettier": "^3.2.5",
    "turbo": "latest"
  },
  "engines": {
    "node": ">=18"
  },
  "packageManager": "yarn@1.22.19",
  "workspaces": [
    "apps/*",
    "packages/*"
  ]
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
